Multiplatform build support

larskanis opened the pull request:
Add support for multiple host platforms.
This way we can provide images for Intel/AMD based as well as ARM based systems.

Image building is now prefixed by the host platform "arm" or "x86":

```
rake build:arm:arm-linux-musl      # Build image for platform arm-linux-musl on linux/arm64
rake build:arm:arm64-darwin        # Build image for platform arm64-darwin on linux/arm64
rake build:arm:jruby               # Build image for JRuby on linux/arm64
```

and:

```
rake build:x86:arm-linux-musl      # Build image for platform arm-linux-musl on linux/amd64
rake build:x86:arm64-darwin        # Build image for platform arm64-darwin on linux/amd64
rake build:x86:jruby               # Build image for JRuby on linux/amd64
```

Only `docker buildx build` supports multiple platforms, so that it is used now, instead of the classic `docker build`.
However buildx doesn't load the images into the local docker registry.
This is why `rake load:images` is added.
It loads all the images for the local platform and makes it available in `docker images`.

The number of images doubles with this patch, so that I get errors building all in parallel.
Therefore the number of parallel tasks should be limited like:

```
rake -j10 build:all
```

flavorjones commented:
IMHO any `buildx build` command we run in this project should pass `--load` to import into docker. This is what I'm doing already in `docker_build_cmd` when `RCD_USE_BUILDX_CACHE` is set.

flavorjones commented:
It might be interesting to note here that I just pushed a github actions pipeline that builds and pushes versioned images, so I no longer have to build releases locally on my (x86_64) development system.

larskanis commented:
> IMHO any `buildx build` command we run in this project should pass `--load` to import into docker.

Hmm, valid point! The `--load` argument is not allowed with multiple architectures. But it makes sense to load at least the local architecture. I changed the rake `build:` tasks to also load the local archs and I removed the dedicated `load:` task.

I adjusted the CI tasks, so that it looks green now. But it is only running the `build:x86` tasks.

Is `masos-14` the only ARM64 arch on github actions? So far there's no linux arm system available, or do you know any? I'll try to get the `macos-14` host system added for building images otherwise.

flavorjones commented:
@larskanis I poked around a bit and it looks like we might be able to apply for free credits as an open-source project: https://github.com/WorksOnArm/GitHub-Runners

Some related references:
- https://www.arm.com/markets/computing-infrastructure/works-on-arm
- https://newsroom.arm.com/blog/github-arm-integration
- https://github.com/actions/partner-runner-images?tab=readme-ov-file

larskanis commented:
OK, I applied for ARM based Github Runners: https://github.com/WorksOnArm/GitHub-Runners/issues/9 . Let's see what they think about it.

For now `macos-14` runners are not helpful for providing arm64 images: #144

larskanis commented:
@flavorjones What do you think about merging this? It allows at least to build and publish the images locally for `x86_64` and `aarch64`.

flavorjones commented:
@larskanis I rebased onto `main` and updated the new CI pipeline for publishing images. If this goes green, let's merge it!

larskanis commented:
Mike, do you have an arm based system locally to create a docker buildx cluster as described in the contributing doc of this PR? I've got a new quite powerful laptop due to fundraiser on https://rubyinstaller.org . 

To my understanding, per platform images can not be simply appended after pushing. Either the two host platforms must be pushed by one buildx command or two images must be pushed separately with different names and then a manifest with links to them must be pushed. 

The rake task `release:images` uses the first method. So we can not release x86 from CI and arm from local that way.

flavorjones commented:
@larskanis Unfortunately, I do not own ARM hardware. Until we figure out how to deploy, we need to maintain the ability to ship the x86 platform images, which is why I rebased and updated the CI pipeline.

flavorjones commented:
I'm OK merging this. Let's continue the conversation about how to deploy these images in https://github.com/rake-compiler/rake-compiler-dock/issues/78?

larskanis commented:
The building of multi-platform images works probably in CI, but publishing the images as snapshot releases doesn't work on a pull request. This is because the cache entries of a pull request are not accessible from another workflow. See the [description here](https://docs.github.com/en/actions/reference/dependency-caching-reference#restrictions-for-accessing-a-cache). I therefore merge this PR like it is and adjust the "publish" and "release" workflows afterwards, when they can access the build cache.
